### PR TITLE
Protect against failed point inversions due to interface kernels

### LINKED ISSUE #18175: Displaced face objects interfere with fake side neighbors (czm)
## Bug Description
As shared with @arovinelli, if you have `use_displaced_mesh = true` in a face postprocessor, then `_reinit_displaced_face` is `true` in `FEProblemBase::reinitNeighbor` and so then call the displaced problem `reinitNeighbor` which produces failed point inversions when interface neighbors don't actually share a common face *and* when the option to use the undisplaced reference points has not been selected. If there is an interface kernel that suggests that the undisplaced reference points should be used *but* it doesn't actually execute on the displaced mesh, then its suggestion is ignored (and I think that's an ok default behavior). A fix would be to introduce a `_reinit_displaced_interface` member that would be `false` if there are no displaced interface kernels, which would avoid reinit of displaced elements around the interface altogether.

## Steps to Reproduce
This input will trigger the failed point inversions warning in `dbg` mode:
```
#
# Stretch + rotation test
#
# This test is designed to compute a uniaxial stress and then follow it as the mesh is rotated by 45 degrees.
#
# The mesh is composed of two, single-elemnt blocks
# The large deforamtion traction separation kinematic assumes linear rotations and uses the velocity gradient L to keep track of area changes, hence it converges to the proper solutoin in the limit of dt->0. Smaller the time step higher the accuracy.

[Mesh]
  [./msh]
  type = GeneratedMeshGenerator
  dim = 3
  nx = 1
  ny = 1
  nz = 2
  xmin = -0.5
  xmax = 0.5
  ymin = -0.5
  ymax = 0.5
  zmin = -1
  zmax = 1
  []
  [./new_block]
    type = SubdomainBoundingBoxGenerator
    input = msh
    block_id = 1
    bottom_left = '-0.5 -0.5 0'
    top_right = '0.5 0.5 0.5'
  []
  [./split]
    type = BreakMeshByBlockGenerator
    input = new_block
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Functions]
  [./stretch]
    type = PiecewiseLinear
    x = '0 1'
    y = '0 1e2'
  [../]
[]

[BCs]
  [./fix_x]
    type = DirichletBC
    preset = true
    value = 0.0
    boundary = left
    variable = disp_x
  [../]
  [./fix_y]
    type = DirichletBC
    preset = true
    value = 0.0
    boundary = bottom
    variable = disp_y
  [../]
  [./fix_z]
    type = DirichletBC
    preset = true
    value = 0.0
    boundary = back
    variable = disp_z
  [../]
  [./back_z]
    type = FunctionNeumannBC
    boundary = front
    variable = disp_z
    use_displaced_mesh = false
    function = stretch
  [../]

[]


[Modules]
  [./TensorMechanics]
    [./Master]
      [./all]
        strain = FINITE
        add_variables = true
        use_finite_deform_jacobian = true
        use_automatic_differentiation = true
        generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_xz'
      [../]
    [../]
  [../]
[]

[Modules/TensorMechanics/CohesiveZoneMaster]
  [./czm1]
    boundary = 'interface'
    displacements = 'disp_x disp_y disp_z'
  [../]
[]

[Materials]
  [./stress]
    type = ADComputeFiniteStrainElasticStress
  [../]
  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 1e3
    poissons_ratio = 0.3
  [../]
  [./czm_3dc]
    type = SalehaniIrani3DCTraction
    boundary = 'interface'
    normal_gap_at_maximum_normal_traction = 0.01
    tangential_gap_at_maximum_shear_traction = 0.005
    maximum_normal_traction = 1000
    maximum_shear_traction = 700
    displacements = 'disp_x disp_y disp_z'
  [../]
[]


[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Postprocessors]
  [./Area]
     type = AreaPostprocessor
     boundary = interface
     use_displaced_mesh = true
  [../]
[]

[Executioner]
  # Executioner
  type = Transient

  solve_type = 'NEWTON'
  line_search = none
  petsc_options_iname = '-pc_type '
  petsc_options_value = 'lu'
  nl_rel_tol = 1e-30
  nl_abs_tol = 1e-10
  l_max_its = 20
  start_time = 0.0
  dt = 0.25
  end_time = 0.25
[]

[Outputs]
  [./out]
    type = CSV
  [../]
[]
```

## Impact
Failed point inversions are always a scary thing. In this case it shouldn't actually impact the simulation, but it does indicate that we are doing unnecessary work.

### PR DESCRIPTION
**Pull request by lindsayad**
Or at least we should not try to invert physical points related to
interface kernels unless we actually have a displaced interface kernel
in our system

Closes #18175

### CODE REVIEW

**Review by arovinelli — DISMISSED** [review]
Seems ok to me, and solves my issue

File: framework/doc/content/syntax/InterfaceKernels/index.md [hunk 1]
```diff
@@ -21,6 +21,18 @@ Interface kernels can be used to provide any general flux condition at an interf
 - `neighbor_var`: The "secondary" variable. This could be $c_1$ from our example above.
 - `boundary`: The interfacial boundary between the subdomains. Note that this must be a sideset and again must exist on the same subdomain as the primary variable. The fact that this boundary is a sideset allows access to variable gradients.
 
+An important private parameter relevant to displaced mesh calculations is
```
Comment (loganharbour) [inline framework/doc/content/syntax/InterfaceKernels/index.md:24]: Hm - just to make sure. There's no case where we would want this to be a public param, correct?

File: test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i [hunk 2]
```diff
@@ -0,0 +1,139 @@
+[Mesh]
```
Comment (loganharbour) [inline test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i:1]: You've got some mismatched hit styles here - make it consistent?

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -3494,9 +3502,11 @@ FEProblemBase::addUserObject(const std::string & user_object_name,
     {
       if (euo || nuo)
         _reinit_displaced_elem = true;
-      else if (suo || iuo)
+      else if (suo)
         // shouldn't we add isuo
         _reinit_displaced_face = true;
+      else if (iuo)
+        _reinit_displaced_neighbor = true;
```
Comment (loganharbour) [inline framework/src/problems/FEProblemBase.C:3509]: 👍🏼

File: framework/doc/content/syntax/InterfaceKernels/index.md (hunk 1 above)
Comment (lindsayad) [reply framework/doc/content/syntax/InterfaceKernels/index.md:24]: This parameter was only introduced to support cohesive zone and its fake neighbors. I think the day when we make fake neighbors a first class capability in libMesh (e.g. they're actually accessible via calls like `neighbor_ptr` and are thus no longer "fake") is the day that we can make this parameter public. As of right now the way we do cohesive zone is kind of an abuse

File: test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i (hunk 2 above)
Comment (lindsayad) [reply test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i:1]: 👍

File: test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+[]
+
+[Variables]
+  [u]
+    order = FIRST
+    family = LAGRANGE
+    block = '0'
+  []
+
+
```
Comment (fdkong) [inline test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i:49]: Do not need many spaces

**Review by fdkong — DISMISSED** [review]
Looks good

File: test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i (hunk 4 above)
Comment (lindsayad) [reply test/tests/interfacekernels/1d_interface/no-failed-point-inversions.i]: Ok I removed the space 😄 Can you re-approve?

**Review by loganharbour — APPROVED** [review]

**Comment by loganharbour** [comment]
This broke next: https://civet.inl.gov/event/70823/ 

Just look at the parallel and recover failures, rest are unrelated)

**Comment by lindsayad** [comment]
Ok I will look at this first thing tomorrow morning

**Pull request merged**